src/charm.py

```diff
@@ -39,10 +41,124 @@ def __init__(self, *args):
         self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.get_initial_password_action, self._on_get_initial_password)
         self._cluster = Patroni(self._unit_ip)
 
+    def _on_get_primary(self, event: ActionEvent) -> None:
+        """Get primary instance."""
+        try:
+            primary = self._patroni.get_primary(unit_name_pattern=True)
+            event.set_results({"primary": primary})
+        except RetryError as e:
+            logger.error(f"failed to get primary with error {e}")
+
+    @retry(
+        stop=stop_after_delay(60),
+        wait=wait_fixed(5),
+        reraise=True,
+    )
+    def _change_primary(self) -> None:
+        """Change the primary member of the cluster."""
+        # Inform the first of the remaining available members to not incur the risk
+        # of triggering a switchover to a member that is also being removed.
+        if not self._patroni.are_all_members_ready():
+            logger.info("could not switchover because not all members are ready")
+            return
+
+        # Try switchover and raise and exception if it doesn't succeed.
+        # If it doesn't happen on time, Patroni will automatically run a fail-over.
+        try:
+            self._patroni.switchover()
+            logger.info("successful switchover")
+        except SwitchoverFailedError as e:
+            logger.error(f"switchover failed with reason: {e}")
+
+    def _get_ip_by_unit(self, unit: Unit) -> str:
+        """Get the IP address of a specific unit."""
+        return self._peers.data[unit].get("private-address")
+
+    @property
+    def _hosts(self) -> set:
+        """List of the current Juju hosts.
+
+        Returns:
+            a set containing the current Juju hosts
+                with the names in the k8s pod name format
+        """
+        peers = self.model.get_relation(PEER)
+        hosts = [self.unit.name.replace("/", "-")] + [
+            unit.name.replace("/", "-") for unit in peers.units
+        ]
+        return set(hosts)
+
+    @property
+    def _patroni(self) -> Patroni:
+        """Returns an instance of the Patroni object."""
+        return Patroni(
+            self._unit_ip,
+            self._storage_path,
+            self._cluster_name,
+            self._member_name,
+            self.app.planned_units(),
+            self._peers_ips,
+            self._get_postgres_password(),
+            self._replication_password,
+        )
+
+    @property
+    def _peers_ips(self) -> List[str]:
+        """Fetch current list of peers IPs.
+
+        Returns:
+            A list of peers addresses (strings).
+        """
+        # Get all members IPs and remove the current unit IP from the list.
+        addresses = self.members_ips
+        current_unit_ip = self._unit_ip
+        if current_unit_ip in addresses:
+            addresses.remove(current_unit_ip)
+        return addresses
```

Comment:
if the function `update_member_ips` hasn't been called the data in `members_ips` might not include all the peers, to be certain that you get the peer ips. I'd suggest doing 

```suggestion
    def _peers_ips(self) -> List[str]:
        """Fetch current list of peers IPs.

        Returns:
            A list of peers addresses (strings).
        """
        # Get all members IPs and remove the current unit IP from the list.
        peer_ips = [ self._get_ip_by_unit(unit) unit in self._peers.units] 
        return peer_ips
```

Reply:
I changed the name of the method and its description to make it more clear. Thanks for pointing this Mia. Changed on https://github.com/canonical/postgresql-operator/pull/10/commits/0d03e9f03e3cb51d1e2cd7353ff315489c110b3f. It is needed to get only the ips of the units that are part of the cluster (the current members).